Documentation/Evolution/StronglyTypedCaptures.md

```diff
@@ -384,147 +426,79 @@ print(match.group(1))
-        }
-    }
-
-// Or with named captures:
-let number = line
-    .firstMatch(of: /(?<binary>[01]+)|(?<decimal>[0-9]+)|(?<hex>[0-9A-F]+)/)
-    .flatMap { (_, n) in
-        switch n {
-        case let .binary(str):
-            return Int(str, radix: 2)
-        case let .decimal(str):
-            return Int(str, radix: 10)
-        case let .hex(str):
-            return Int(str, radix: 16)
-        }
-    }
-```
-
-In the fullness of time, we'd like to design a language feature to support this.
-In the meantime, we would like to do the best we can and leave the door open for
-a source-compatible migration.
-
-With variadic generics we think we can define the following `Alternation` type.
+// Equivalent result builder syntax:
+//     let numberAlternationRegex = Pattern {
+//         OneOf {
+//             OneOrMore(.binaryDigit).capture()
+//             OneOrMore(.decimalDigit)
+//             OneOrMore(.hexDigit).capture()
+//         }
+//     }
+//     .flattened()
 
-```swift
-@dynamicMemberLookup
-public struct Alternation<Captures> { ... }
-
-extension<Option...> Alternation where Captures == (Option...) {
-    public var options: (Option?...) { get }
-    
-    public subscript<T>(dynamicMember keyPath: KeyPath<(Option?...), T>) -> T {
-      options[keyPath: keyPath]
-    }
-}
-```
+let scalarRangeAlternation = /([0-9a-fA-F]+)\.\.([0-9a-fA-F]+)|([0-9a-fA-F]+)/
+// Positions in result:     1 ^~~~~~~~~~~~~~  2 ^~~~~~~~~~~~~~
+//                                                           3 ^~~~~~~~~~~~~~
+// => `Regex<(Substring, Substring?, Substring?, Substring?)>
 
-An optional projection property, `options`, presents all options as a tuple of
-optionals. The Standard Library will provide the runtime guarantee that only one
-element of the `options` tuple will be non-nil. The programmer can use a
-`switch` statement to pattern-match the tuple and handle each case, or directly
-access the tuple's properties via key path dynamic member lookup.
+// Equivalent result builder syntax:
+//     let scalarRangeAlternation = Pattern {
+//         OneOf {
+//             Group {
+//                 OneOrMore(.hexDigit).capture()
+//                 ".."
+//                 OneOrMore(.hexDigit).capture()
+//             }
+//             OneOrMore(.hexDigit).capture()
+//         }
+//     }
+//     .flattened()
 
-```swift
-let number = line
-    .firstMatch(of: /([01]+)|([0-9]+)|([0-9A-F]+)/)
-    .flatMap { (_, n) in
-        switch n.options {
-        case let (binary?, nil, nil):
-            return Int(binary, radix: 2)
-        case let (nil, decimal?, nil):
-            return Int(decimal, radix: 10)
-        case let (nil, nil, hex?):
-            return Int(hex, radix: 16)
-        default:
-            fatalError("unreachable")
-        }
-    }
+let nestedScalarRangeAlternation = /(([0-9a-fA-F]+)\.\.([0-9a-fA-F]+))|([0-9a-fA-F]+)/
+// Positions in result:           1 ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~   
```

Comment:
It would be clearer if we add a line about `0`-th match because down below we show its generic type.  Same suggestions for other occurrences.